### PR TITLE
Test for corner-cases in mul_overflow.pass.cpp

### LINKED ISSUE #7739: [FEA]: Optimize `cuda::mul_overflow`
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

libcu++

### Is your feature request related to a problem? Please describe.

In #4415, we added `cuda::mul_overflow`. Currently, we upcast/use `cuda::mul_hi` for the implementation. We should find a way to optimize it.

### Describe the solution you'd like

Device optimized `cuda::mul_overflow`

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by acosmicflamingo**
## Description

While working on optimizing `mul_overflow` kernel to resolve #7739, I have been using a local test-suite to catch corner-cases to ensure I don't introduce any regressions. For some odd reason, I am finding that when I added the tests to `mul_overflow.pass.cpp` in the other PR, I'd see an atrocious execution time:
```bash
--------------------------------------------------------------------------
94.41s: libcu++ :: cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp
4.89s: libcu++ :: std/algorithms/alg.nonmodifying/alg.count/pstl.count_if.pass.cpp
4.88s: libcu++ :: std/algorithms/alg.nonmodifying/alg.for_each/pstl.for_each.pass.cpp
4.81s: libcu++ :: std/algorithms/alg.nonmodifying/alg.for_each/pstl.for_each_n.pass.cpp
4.76s: libcu++ :: std/algorithms/alg.nonmodifying/alg.all_of/pstl.all_of.pass.cpp
4.70s: libcu++ :: std/numerics/numeric.ops/reduce/pstl.reduce.pass.cpp
4.70s: libcu++ :: std/algorithms/alg.nonmodifying/alg.find/pstl.find_if.pass.cpp
4.49s: libcu++ :: std/algorithms/alg.nonmodifying/alg.find/pstl.find.pass.cpp
3.54s: libcu++ :: std/algorithms/alg.nonmodifying/alg.count/pstl.count.pass.cpp
3.50s: libcu++ :: std/algorithms/alg.nonmodifying/alg.find/pstl.find_if_not.pass.cpp
3.37s: libcu++ :: std/algorithms/alg.nonmodifying/alg.none_of/pstl.none_of.pass.cpp
3.26s: libcu++ :: std/algorithms/alg.nonmodifying/alg.any_of/pstl.any_of.pass.cpp
2.47s: libcu++ :: cuda/containers/views/mdspan/default_accessor/accessor.relation.fail.cpp
2.30s: libcu++ :: std/utilities/utility/pairs/pair.piecewise/piecewise_construct_t.fail.cpp
2.21s: libcu++ :: std/utilities/meta/meta.trans/meta.trans.other/result_of.fail.cpp
2.15s: libcu++ :: std/text/format/format.fmt.string/ctor_fails/3.fail.cpp
2.15s: libcu++ :: std/text/format/format.fmt.string/ctor_fails/4.fail.cpp
2.13s: libcu++ :: std/algorithms/alg.sorting/alg.sort/sort/sort_constexpr.pass.cpp
2.06s: libcu++ :: std/text/format/format.fmt.string/ctor_fails/1.fail.cpp
2.05s: libcu++ :: std/iterators/iterator.container/empty.container.fail.cpp
```

This PR (and CI run results) would assure me that the slow times are truly local in scope, while making sure that the current version of `mul_overflow` covers all corner-cases.

I also added `test_corner_cases` below `test_exhaustive` since I didn't want to have it called repeatedly (especially because I'm explicitly controlling types to test the corner cases).

## Checklist
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by acosmicflamingo** [comment]
@fbusato if you agree with the approach, can you run `/ok to test` so we can see what the CI output is?

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 1]
```diff
@@ -92,17 +92,71 @@ TEST_FUNC constexpr bool test_type()
...
+  constexpr auto uint_max = cuda::std::numeric_limits<unsigned>::max();
+
+  // 1. Boundary edge-cases
+  test_mul_overflow<int, int, int>(int_min, 0, 0, false);
+  test_mul_overflow<int, int, int>(0, int_max, 0, false);
+  test_mul_overflow<int, int, int>(-1, int_min, int_min, true);
+  test_mul_overflow<int, int, int>(int_min, -1, int_min, true);
+  test_mul_overflow<int, int, int>(int_min, int_min, 0, true);
+  test_mul_overflow<unsigned int, int, int>(2, int_max, static_cast<uint32_t>(int_max) * 2u, false);
+  test_mul_overflow<unsigned int, int, int>(uint_max, 2u, uint_max - 1, true);
```
[the remark this answers is not in the document]
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Will be resolved in upcoming commit

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 2]
```diff
@@ -92,17 +92,71 @@ TEST_FUNC constexpr bool test_type()
...
+    static_cast<__uint128_t>(1) << 100, static_cast<__uint128_t>(1) << 100, static_cast<__uint128_t>(0), true);
+  test_mul_overflow<__uint128_t, __uint128_t, __uint128_t>(
+    uint128_max, static_cast<__uint128_t>(2), uint128_max - 1, true);
+  test_mul_overflow<__uint128_t, __uint128_t, __uint128_t>(
+    uint128_max, static_cast<__uint128_t>(0), static_cast<__uint128_t>(0), false);
+  test_mul_overflow<__uint128_t, long long, __uint128_t>(
+    ~0ull,
+    static_cast<__uint128_t>(5) << 100,
+    (static_cast<__uint128_t>(0xffffffb000000000ULL) << 64) | static_cast<__uint128_t>(0),
+    false);
```
[the remark this answers is not in the document]
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Will be resolved in upcoming commit

**Comment by acosmicflamingo** [comment]
Although the test failures themselves do not seem to be related to this PR, I actually found that there _is_ a slowdown! This is from [libcu++ nvcc GCC / [CTK13.3 GCC15 C++17] Build(amd64)](https://github.com/NVIDIA/cccl/actions/runs/29130796447/job/86485683082?pr=9792#logs) log:
```
2026-07-11T00:08:31.4070895Z 5 warning(s) in tests
2026-07-11T00:08:31.4205936Z Slowest Tests:
2026-07-11T00:08:31.4206356Z --------------------------------------------------------------------------
2026-07-11T00:08:31.4206839Z 139.30s: libcu++ :: cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp
2026-07-11T00:08:31.4207250Z 15.07s: libcu++ :: cuda/argument/argument_traits.pass.cpp
...
```

However, with other runs I have noticed that the slowest tests very quite a bit. Here are two random ones I pulled up:
```
# From https://github.com/NVIDIA/cccl/pull/8740
2026-07-06T23:54:32.4247194Z 5 warning(s) in tests
2026-07-06T23:54:32.4364490Z Slowest Tests:
2026-07-06T23:54:32.4364889Z --------------------------------------------------------------------------
2026-07-06T23:54:32.4365397Z 46.15s: libcu++ :: std/algorithms/alg.nonmodifying/alg.any_of/pstl.any_of.pass.cpp
2026-07-06T23:54:32.4365933Z 45.16s: libcu++ :: std/algorithms/alg.nonmodifying/alg.for_each/pstl.for_each_n.pass.cpp
2026-07-06T23:54:32.4366554Z 44.37s: libcu++ :: std/algorithms/alg.nonmodifying/alg.count/pstl.count.pass.cpp
2026-07-06T23:54:32.4367040Z 42.52s: libcu++ :: std/algorithms/alg.nonmodifying/alg.for_each/pstl.for_each.pass.cpp
2026-07-06T23:54:32.4367542Z 42.49s: libcu++ :: std/algorithms/alg.nonmodifying/alg.none_of/pstl.none_of.pass.cpp
2026-07-06T23:54:32.4368002Z 42.29s: libcu++ :: std/algorithms/alg.nonmodifying/alg.find/pstl.find.pass.cpp
2026-07-06T23:54:32.4368471Z 41.71s: libcu++ :: std/algorithms/alg.nonmodifying/alg.find/pstl.find_if_not.pass.cpp

# From https://github.com/NVIDIA/cccl/pull/9694
2026-07-02T16:19:45.3816915Z 5 warning(s) in tests
2026-07-02T16:19:45.3932122Z Slowest Tests:
2026-07-02T16:19:45.3932481Z --------------------------------------------------------------------------
2026-07-02T16:19:45.3933312Z 274.54s: libcu++ :: std/algorithms/alg.modifying/alg.copy/copy_ptr.pass.cpp
2026-07-02T16:19:45.3933821Z 229.07s: libcu++ :: std/algorithms/alg.modifying/alg.move/move.pass.cpp
2026-07-02T16:19:45.3934254Z 218.02s: libcu++ :: std/algorithms/alg.modifying/alg.move/move_backward.pass.cpp
2026-07-02T16:19:45.3934768Z 201.21s: libcu++ :: cuda/pipeline/pipeline_group_concept_thread_scope_block.pass.cpp
2026-07-02T16:19:45.3935233Z 174.62s: libcu++ :: std/algorithms/alg.modifying/alg.copy/copy_backward.pass.cpp
2026-07-02T16:19:45.3935689Z 146.03s: libcu++ :: cuda/pipeline/pipeline_group_concept_thread_scope_device.pass.cpp
2026-07-02T16:19:45.3936147Z 145.12s: libcu++ :: cuda/numeric/overflow.arithmetic/add_overflow.pass.cpp
2026-07-02T16:19:45.3936583Z 136.65s: libcu++ :: std/algorithms/alg.modifying/alg.generate/generate_n.pass.cpp
2026-07-02T16:19:45.3937046Z 135.44s: libcu++ :: cuda/pipeline/pipeline_group_concept_thread_scope_system.pass.cpp
2026-07-02T16:19:45.3937499Z 132.23s: libcu++ :: cuda/numeric/overflow.arithmetic/sub_overflow.pass.cpp
2026-07-02T16:19:45.3937904Z 125.03s: libcu++ :: cuda/memcpy_async/memcpy_async_large.pass.cpp
2026-07-02T16:19:45.3938325Z 124.49s: libcu++ :: cuda/pipeline/pipeline_memcpy_async_thread_scope_block_32.pass.cpp
```

**Comment by acosmicflamingo** [comment]
Great news; I am finding that this behavior might actually be completely expected, and that the kernels are being cached. I can deterministically get the slowdown by making sure the compiler has not come across the `.cpp` file in a particular state before. For example, merely copying and pasting a line and flipping the operands will cause a slowdown in the first run.

With that being said, merging this now would make the `mul_overflow` optimization PR easier to sift through, but I'm fine with it being merged there if y'all want 🙇🏽‍♂️

**Comment by fbusato** [comment]
smaller PRs are always better. My current concern is that the new test cases are hardcode for specific types. Would be possible to generalize them? e.g.
```cpp
test_mul_overflow<int32_t, int32_t, int32_t>(int_min, 0, 0, false);
```
to 
```cpp
test_mul_overflow<T, T, T>(min_value, T{0}, T{0}, false);
```
also, please check that they are not already covered by previous tests.

**Comment by acosmicflamingo** [comment]
@fbusato hm, if I can still catch the same kind of errors then I think it'd be a good idea. I hardcoded it (even going so far as using std types instead of more generic types whose bitwidth could vary across platforms) because I wanted to be certain that the operand bit sizes were exactly a certain width, even if it came at the expense of readability (they are corner-case tests after all hahah). Let's see what I can whip up!

Edit: I'll double-check that there isn't any re-testing going on too. I am only 99% sure since I had kept avoid duplication in mind, but it never hurts to verify :)

**Comment by acosmicflamingo** [comment]
@fbusato you know...now that we found that the increase in test run-time was not because of anything inherent in the implementation, let me see whether I could add some of these corner-cases within the section of code that utilize generics. Perhaps I could strike a balance between checking corner-cases without greatly increasing the time it takes to even compile the `mul_overflow.pass.cpp` file.

**Comment by acosmicflamingo** [comment]
@fbusato I remembered why I avoided adding the corner-cases along with the rest of the checks in `test_type` and using generics; it becomes quite a mess simultaneously making sure everything works with static type-checking during compilation and ensuring that I'm asserting what I originally intended to test.

Is the problem with using explicit types in this PR due to how much harder it becomes to read the code, or is it a preference for generics over explicit std types altogether (or wanting to include everything within test_type)?

Perhaps taking an approach like this would be what you were asking for (although the generic type pile up when I have combinations of signed and unsigned bits with 32/64 bit combinations):
```C++
// 1. Boundary edge-cases
using T1 = cuda::std::int32_t;
using U1 = cuda::std::int32_t;
test_mul_overflow<T1, T1, T1>(int_min, int_min, T1{0}, true);
test_mul_overflow<U1, T1, T1>(2, int_max, U1{int_max} * 2u, false);
test_mul_overflow<U1, U1, U1>(uint_max, 2ull, uint_max - 1ull, true);
```

**Comment by acosmicflamingo** [comment]
Latest commit uses generics to clean up lots of the code (thanks @fbusato for the suggestion!). I also merged the latest version of `main` so the unrelated `simd` test failures should be gone (assuming no issues were recently introduced)

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 3]
```diff
@@ -92,17 +93,82 @@ TEST_FUNC constexpr bool test_type()
...
+  {
+    using T             = cuda::std::int32_t;
+    using T2x           = cuda::std::int64_t;
+    using U             = cuda::std::uint32_t;
+    using U2x           = cuda::std::uint64_t;
+    constexpr auto min  = cuda::std::numeric_limits<T>::min();
+    constexpr auto umax = cuda::std::numeric_limits<U>::max();
+    test_mul_overflow<T2x, T, T>(min, -1, T2x{min} * (-1ll), false);
+    test_mul_overflow<T2x, T, T>(min, min, T2x{min} * T2x{min}, false);
+    test_mul_overflow<T2x, T, T>(T{umax}, T{umax}, T2x{umax * umax}, false);
```
[the remark this answers is not in the document]
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Although the test does compile, I agree that the code does read as if a typo exists even though it's what I intended. This will be addressed in an upcoming commit

**Comment by acosmicflamingo** [comment]
> ## 😬 CI Workflow Results
> ### 🟥 Finished in 2h 56m: Pass: 98%/71 | Total: 1d 05h | Max: 1h 32m | Hits: 99%/328970
> See results [here](https://github.com/NVIDIA/cccl/actions/runs/29373234766).

I confirmed (specifically looking into https://github.com/NVIDIA/cccl/pull/9853) that the errors are unrelated

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 4]
```diff
@@ -92,17 +93,83 @@ TEST_FUNC constexpr bool test_type()
...
+  {
+    using T             = cuda::std::int32_t;
+    using U             = cuda::std::uint32_t;
+    constexpr auto max  = cuda::std::numeric_limits<T>::max();
+    constexpr auto min  = cuda::std::numeric_limits<T>::min();
+    constexpr auto umax = cuda::std::numeric_limits<U>::max();
+    test_mul_overflow<T, T, T>(min, min, T{0}, true);
+    test_mul_overflow<U, T, T>(2, max, U{max} * 2u, false);
+    test_mul_overflow<U, U, U>(umax, 2ull, umax - 1ull, true);
+  }
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp:98]: these 3 test cases can be moved in the function above. The constrain could be
```
is_signed_v<L> && is_unsigned_v<R> && sizeof(L) == sizeof(T) && (is_same_v<Res, T> || is_same_v<Res, R>)
```

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 5]
```diff
@@ -92,17 +93,83 @@ TEST_FUNC constexpr bool test_type()
...
+  // 2. Explicit wider Result type
+  {
+    using T             = cuda::std::int32_t;
+    using T2x           = cuda::std::int64_t;
+    using U             = cuda::std::uint32_t;
+    using U2x           = cuda::std::uint64_t;
+    constexpr auto min  = cuda::std::numeric_limits<T>::min();
+    constexpr auto max  = cuda::std::numeric_limits<T>::max();
+    constexpr auto umax = cuda::std::numeric_limits<U>::max();
+    test_mul_overflow<T2x, T, T>(min, -1, T2x{min} * (-1ll), false);
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: this case looks already covered

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 6]
```diff
@@ -92,17 +93,83 @@ TEST_FUNC constexpr bool test_type()
...
+    test_mul_overflow<T2x, T, T>(min, min, T2x{min} * T2x{min}, false);
+    test_mul_overflow<T2x, T, T>(max, max, T2x{max} * T2x{max}, false);
+    test_mul_overflow<U2x, U, U>(umax, umax, U2x{umax} * U2x{umax}, false);
+    test_mul_overflow<T2x, U, U>(umax, umax, -8589934591, true);
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: I think all these cases can be generalized as well

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 7]
```diff
@@ -92,17 +93,83 @@ TEST_FUNC constexpr bool test_type()
...
+  }
+
+  // 3. Both operands negative, large magnitude (non-overflow and overflow)
+  {
+    using T = cuda::std::int32_t;
+    test_mul_overflow<T, T, T>(-40000, -50000, 2000000000, false);
+    test_mul_overflow<T, T, T>(-50000, -50000, -1794967296, true);
+  }
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: this can fall in "// 4. Test other numbers" when it is representable by T

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 8]
```diff
@@ -92,17 +93,83 @@ TEST_FUNC constexpr bool test_type()
...
+
+  // 4. Overflow from downcasting
+  {
+    using T   = cuda::std::int8_t;
+    using T4x = cuda::std::int32_t;
+    using U4x = cuda::std::uint32_t;
+    test_mul_overflow<T, T4x, T4x>(1000, 1000, T{64}, true);
+    test_mul_overflow<T, U4x, T>(T{17}, U4x{14}, T{-18}, true);
+  }
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: same here

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 9]
```diff
@@ -92,17 +93,83 @@ TEST_FUNC constexpr bool test_type()
...
+
+    test_mul_overflow<U2x, U2x, U2x>(3, 4, 12, false);
+    test_mul_overflow<U2x, U2x, U2x>(
+      U2x{~0ull}, // 2^64 - 1
+      U2x{1ull} << 63, // 2^63
+      (U2x{0x7fffffffffffffffULL} << 64) | U2x{0x8000000000000000ULL},
+      false);
+    test_mul_overflow<U2x, U2x, U2x>(U2x{1} << 100, U2x{1} << 100, U2x{0}, true);
+    test_mul_overflow<U2x, U2x, U2x>(umax2x, U2x{2}, umax2x - 1, true);
+    test_mul_overflow<U2x, U2x, U2x>(umax2x, U2x{0}, U2x{0}, false);
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: these two look generic as well

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 10]
```diff
@@ -92,17 +93,83 @@ TEST_FUNC constexpr bool test_type()
...
+  }
+
+#if _CCCL_HAS_INT128()
+  // 5. __uint128_t
+  {
+    using U               = cuda::std::uint64_t;
+    using U2x             = __uint128_t;
+    constexpr auto umax2x = cuda::std::numeric_limits<U2x>::max();
+
+    test_mul_overflow<U2x, U2x, U2x>(3, 4, 12, false);
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: this can be also part of `// 4. Test other numbers`

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp (hunk 5 above)
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Hm, yes. Functionally, it is covered within `// 3. Test T(-1) * T_MIN case`. Looking at it now, I don't know whether I was reluctant to have the test depend on the output of `cuda::uabs`, which is silly since CI would complain about it far before it would pose a problem here. I'll go ahead and remove line 122.

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp (hunk 7 above)
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: It does. I'll remove the test in 132 since what I really cared about was 133 (which I'll also move to the section you mentioned)

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp (hunk 10 above)
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Very true. I'll move it there.

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp (hunk 4 above)
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp:98]: Hm, I suppose if we're going to go the route of generalizing these corner-cases, I could scrap using those specific types (e.g. `T`, `T2x`, `U4x`) altogether and lean on constexpr instead.

Something like this had crossed my mind earlier:
```diff
   test_mul_overflow<Res, L, R>(1, -1, -1);
   test_mul_overflow<Res, L, R>(-1, -1, 1);
 
-  // 3. Test T(-1) * T_MIN case
+  // 3. Boundary edge-cases
   if constexpr (cuda::std::is_signed_v<L> && cuda::std::is_signed_v<R>)
   {
     // nvcc 12.0 seems to have some issues during constant evaluation of this test. Works fine with later versions, so
     // let's just disable this test.
 #if !_CCCL_CUDA_COMPILER(NVCC, ==, 12, 0)
+    // Test T(-1) * T_MIN case
     test_mul_overflow<Res, L, R>(
       cuda::std::numeric_limits<L>::min(), -1, cuda::uabs(cuda::std::numeric_limits<L>::min()));
     test_mul_overflow<Res, L, R>(
       -1, cuda::std::numeric_limits<R>::min(), cuda::uabs(cuda::std::numeric_limits<R>::min()));
+
+    // T_MIN * T_MIN
+    if constexpr (cuda::std::is_same_v<L, R> && sizeof(L) >= sizeof(Res))
+    {
+      test_mul_overflow<Res, L, R>(
+        cuda::std::numeric_limits<L>::min(), cuda::std::numeric_limits<L>::min(), Res{0}, true);
+    }
```

I was worried that the test could end up bloated with all these very specific conditional statements, but I'm getting the sense that it'd be preferred to hardcoding types. 

If we want to generalize even more (e.g. remove the size-comparison logic), then what might be best is to take the same approach currently used in `mul_overflow.h` to ensure no overflow occurs when calculating the expected result:
```C++
 if constexpr (sizeof(L) < sizeof(cuda::std::__cccl_uintmax_t) && sizeof(R) < sizeof(cuda::std::__cccl_uintmax_t))
  {
    constexpr auto __max_nbits = cuda::std::max(cuda::std::__num_bits_v<L>, cuda::std::__num_bits_v<R>);
    using _Up = cuda::std::__make_nbit_int_t<2 * __max_nbits, cuda::std::is_signed_v<L> || cuda::std::is_signed_v<R>>;
    // T_MIN * T_MIN
    test_mul_overflow<Res, L, R>(cuda::std::numeric_limits<L>::min(),
                                 cuda::std::numeric_limits<R>::min(),
                                 _Up{cuda::std::numeric_limits<L>::min()} * _Up{cuda::std::numeric_limits<R>::min()});
  }
```

Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp:98]: Pushed a commit that uses this change, assuming that its elegance and the corner-cases it allows to test will be gladly accepted :)

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp (hunk 9 above)
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Done

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp (hunk 8 above)
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Done!

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp (hunk 6 above)
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Yup!

**Comment by acosmicflamingo** [comment]
Corner-cases have been generalized and moved with the other `test_mul_overflow` function calls! I utilized the upcasting approach currently used in `mul_overflow.h` to make it very easy to assert multiplication results of corner cases and not encounter compile-time evaluation errors. Doing so collapses much of what I was trying to assert, so I'm quite happy with how @fbusato 's suggestion turned out.

I think after that commit, the PR will be welcomed with open-arms *fingers crossed*

...and even if for whatever reason there is a case that's missed, I'm now familiar enough with the `mul_overflow.pass.cpp` where I can just update it accordingly when working on the actual `mul_overflow` PR of interest.

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 11]
```diff
@@ -93,6 +95,44 @@ TEST_FUNC constexpr bool test_type()
...
+      test_mul_overflow<Res, L, R>(
+        cuda::std::numeric_limits<__uint128_t>::max(),
+        __uint128_t{2},
+        cuda::std::numeric_limits<__uint128_t>::max() - 1,
+        true);
+    }
+    else if constexpr (cuda::std::is_same_v<L, unsigned long long>)
+    {
+      test_mul_overflow<Res, L, R>(
+        ~0ull, __uint128_t{5} << 100, __uint128_t{0xffffffb000000000ULL} << 64 | __uint128_t{0}, true);
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: I don't think `| __uint128_t{0}` is useful

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp [hunk 12]
```diff
@@ -93,6 +95,44 @@ TEST_FUNC constexpr bool test_type()
       sizeof(L) >= sizeof(Res));
   }
 
+  // 5. Test T_MIN * T_MIN and T_MAX * T_MAX
+  if constexpr (sizeof(L) < sizeof(cuda::std::__cccl_uintmax_t) && sizeof(R) < sizeof(cuda::std::__cccl_uintmax_t))
+  {
+    constexpr auto __max_nbits = cuda::std::max(cuda::std::__num_bits_v<L>, cuda::std::__num_bits_v<R>);
+    using _Up = cuda::std::__make_nbit_int_t<2 * __max_nbits, cuda::std::is_signed_v<L> || cuda::std::is_signed_v<R>>;
+    test_mul_overflow<Res, L, R>(cuda::std::numeric_limits<L>::min(),
```
Comment (fbusato) [inline libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: nit. could you please create a variable for `cuda::std::numeric_limits<type>::min/max()` to make the code more readable?  same for __uint128_t case.

**Review by fbusato — APPROVED** [review]
looks good. just a couple of very minor issue

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp (hunk 12 above)
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Sure! I actually took this approach originally, but noticed this block earlier in the code:
```C++
#if !_CCCL_CUDA_COMPILER(NVCC, ==, 12, 0)
    test_mul_overflow<Res, L, R>(
      cuda::std::numeric_limits<L>::min(), -1, cuda::uabs(cuda::std::numeric_limits<L>::min()));
    test_mul_overflow<Res, L, R>(
      -1, cuda::std::numeric_limits<R>::min(), cuda::uabs(cuda::std::numeric_limits<R>::min()));
#endif // !_CCCL_CUDA_COMPILER(NVCC, ==, 12, 0)
```
I thought perhaps it was written that way so the reader doesn't have to scroll somewhere to see what the value actually is. would you like that I apply your suggestion to those areas as well?

Edit: I'll default to do it everywhere since it really does make the code much more readable

File: libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp (hunk 11 above)
Comment (acosmicflamingo) [reply libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp]: Hah, your instincts are correct. I should also actually leverage test_mul_overflow handling the casting for me; I'll be polishing other parts in addition to removing that.

**Comment by acosmicflamingo** [comment]
@fbusato good eye! Happy to tweak those minor issues. fyi I removed the conditional pragmas (you'll see it in the diff) since the constant evaluation is now occurring elsewhere, and I figure the function calls should now work (and if not, I'll revert the commit). 

Perhaps if you could please run the test suite once more, we can get a confirmation that everything truly is good now (works locally, but always nice to see 🟢 in CI)

**Comment by acosmicflamingo** [comment]
Don't know why the pre-commit failed...I'm running my code in VSCode and there are hooks that reformat everything any time I save the file. Even checking locally at my diff, I don't see any trailing whitespace.

Edit: I think CI is being a bit silly, reverting today's changes didn't give me a clean pre-commit run

Edit 2: Yes, it's now green after merging with `main` from GitHub and not using local `upstream/main`

**Comment by acosmicflamingo** [comment]
@fbusato could you please run `/ok to test 19ddd9f29ca4861a6f0a578d638705e6fec4d123` when you have a chance? Hopefully this should be the last time you do so for me (in this PR 😉)

**Comment by acosmicflamingo** [comment]
@fbusato sorry to bother, but would you be able to run `/ok to test 78df2a50d043bf7fb12190d535ae48c7d08c64c8` for yours truly? :)

**Comment by miscco** [comment]
/ok to test [78df2a5](https://github.com/NVIDIA/cccl/pull/9792/commits/78df2a50d043bf7fb12190d535ae48c7d08c64c8)

**Review by miscco — APPROVED** [review]

**Comment by miscco** [comment]
Looks like old MSVC is broken here, can you investigate those?

**Comment by acosmicflamingo** [comment]
@miscco this is quite a unique bug...only impacts `CTK 12.0` on Windows. I don't have a Windows machine with CTK 12.0 installed, so I'm having a very hard time finding a way to locally reproduce this. What a bummer godbolt doesn't support NVCC and Windows either, that could've been helpful here. There also doesn't seem to be any mentions or fixes in any CHANGELOG or release notes that would explain what's going on either.

I could try and push a few commits to see if anything works after you run `/ok to test` for me, but perhaps it'd be a better use of resources if we simply enclose that test within a conditional pragma so it only runs when CTK version is not 12.0? There is precedence for going that route [here](https://github.com/NVIDIA/cccl/blob/main/libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/mul_overflow.pass.cpp#L74) btw (and in fact, the problematic line could be perfect in that block since the two numbers I'm multiplying are both signed)
